[FSDP] Full precision in eval mode

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97645

If model.eval() is true, then runs the model in full precision.

Changes:
- Changed _force_full_precision to check self.is_training
- Check for _force_full_precision when casting gradients to reduced dtype
- Small change when accessing _full_prec_param_padded
- tests for class based and fully_shard APIs

Differential Revision: [D43933690](https://our.internmc.facebook.com/intern/diff/D43933690/)

rohan-varma committed:
[FSDP] Full precision in eval mode

If model.eval() is true, then runs the model in full precision.

Changes:
- Changed _force_full_precision to check self.is_training
- Check for _force_full_precision when casting gradients to reduced dtype
- Small change when accessing _full_prec_param_padded
- tests for class based and fully_shard APIs

Differential Revision: [D43933690](https://our.internmc.facebook.com/intern/diff/D43933690/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Full precision in eval mode"


If model.eval() is true, then runs the model in full precision.

Changes:
- Changed _force_full_precision to check self.is_training
- Check for _force_full_precision when casting gradients to reduced dtype
- Small change when accessing _full_prec_param_padded
- tests for class based and fully_shard APIs

Differential Revision: [D43933690](https://our.internmc.facebook.com/intern/diff/D43933690/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Full precision in eval mode"


If model.eval() is true, then runs the model in full precision.

Changes:
- Changed _force_full_precision to check self.is_training
- Check for _force_full_precision when casting gradients to reduced dtype
- Small change when accessing _full_prec_param_padded
- tests for class based and fully_shard APIs

Differential Revision: [D43933690](https://our.internmc.facebook.com/intern/diff/D43933690/)

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -2065,8 +2066,11 @@ def _uses_reduce_mixed_precision(self) -> bool:
         return (
-            self._training_state == HandleTrainingState.SUMMON_FULL_PARAMS
-            and self._uses_param_mixed_precision
+            (self._uses_param_mixed_precision or self._uses_reduce_mixed_precision) and (
```
This changes the existing semantics when in training mode. I wanted to check that that is intentional.
Do we want to confound forcing parameter full precision with forcing reduce full precision?

From my view, this will unnecessarily take the full precision all-gather code path even when parameters are already in full precision, which is not explicitly bad. I just wanted to hear what your thoughts on this was.

rohan-varma replied:
I needed this change in the case where uses_param_mixed_precision=False but uses_reduced_mixed_precision=True.

In that case, _get_padded_unsharded_flat_param would crash trying to get the `_full_prec_param_padded`, since that is only allocated in parameter mixed precision case. 

Wonder if you have any alternatives here?

rohan-varma replied:
This might help:

1. Yes, we want to force full precision gradient reduction, even  when `use_param_mixed_precision=False` (but `uses_reduced_mixed_precision=True`, indicating user wants to only do reduction in reduced precision). 
2. To enable this, we changed the check in `_force_full_precision` to check if we're used param mixed precision OR reduce mixed precision. Previously if we were only using reduced mixed precision, we'd return False. 
3. Due to this change, _force_full_precision now returns True in `_get_padded_unsharded_flat_param` when doing only reduction mixed precision, when previously it would return False. But, _full_prec_param_padded is only defined when using parameter mixed precision. So in `_get_padded_unsharded_flat_param`, we changed `if self._force_full_precision` -> `if self._force_full_precision and self._uses_param_mixed_precision`.

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -519,6 +525,42 @@ def _root_pre_forward(
+        if len(buffers) > 0 and len(buffer_dtypes_for_computation) > 0:
+            buffer, buffer_dtype_for_computation = buffers[0], buffer_dtypes_for_computation[0]
+            if buffer.dtype != buffer_dtype_for_computation:
+                # Assume we have to cast everything if there is one mismatch
```
Why do we only check the 0th buffer for mismatch?

rohan-varma replied:
good point, moving this to any() check.

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -414,9 +414,15 @@ def _pre_forward(
+            should_cast_forward = False
+            break
+
+    if should_cast_forward and state.mixed_precision.cast_forward_inputs:
```
Is this branch covered by unit tests?

rohan-varma replied:
Yes, if this wasn't added, there would be a dtype mismatch because inputs are in lower precision and parameters are not.

awgu replied:
Just in case: I meant I did not see passing `cast_forward_inputs=True` in the added unit tests.

rohan-varma replied:
good point, this is false by default, will add to tests

awgu reviewed (commented):
This is the first round of higher-level comments.

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -414,9 +414,15 @@ def _pre_forward(
-    # Recursively convert args and kwargs to specified precision.
-    input_dtype: Optional[torch.dtype] = state.mixed_precision.param_dtype
-    if state.mixed_precision.cast_forward_inputs:
+    should_cast_forward = True
+    for handle in state._handles:
+        if handle._force_full_precision:
+            should_cast_forward = False
+            break
```
nit: Should we be consistent in naming?
Below you directly say `cast_buffers_to_full_precision` without a preceding `should_`.

This name should probably also mention "inputs", e.g. `should_cast_forward_inputs`. Then, maybe the one below should be `should_cast_buffers_to_full_prec`.

---

```suggestion
    should_cast_forward_inputs = all(not handle._force_full_precision for handle in state._handles)
```
`all()` short circuits (https://docs.python.org/3/library/functions.html#all).

awgu reviewed (commented):
This is the second round of lower-level comments and nits.

rohan-varma commented on test/distributed/fsdp/test_fsdp_mixed_precision.py:
```diff
@@ -235,21 +239,30 @@ class TestFSDPMixedPrecision(FSDPTest):
         )
         return model
 
+    def _get_simple_nested_model_composable(self, param_dtype, run_checks, *fsdp_args, **fsdp_kwargs):
```
I'm not sure if this is the best, but having separate mixed precision tests for composable and wrapper in different files would be tough to keep them in sync. I think we can go ahead with this for now and follow up later as BE work?

awgu replied:
Sure, sounds good to me!

I was thinking whether we wanted some kind of common file with the base version of each test. However, there will be some challenge to unify wrapper and composable always. Either way, we can just be cognizant that some composable tests are in the wrapper test files too.

rohan-varma commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -1190,7 +1190,8 @@ def prepare_gradient_for_optim(self):
 
         def cast_grad_to_param_dtype_if_needed(flat_param):
-            if self._keep_low_precision_grads:
+            # TODO (rohan-varma): test for full precision with keep_low_precision_grads
```
Filed issue: https://github.com/pytorch/pytorch/issues/97670

awgu commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -2065,8 +2066,11 @@ def _uses_reduce_mixed_precision(self) -> bool:
+            (self._uses_param_mixed_precision or self._uses_reduce_mixed_precision) and (
+                self._training_state == HandleTrainingState.SUMMON_FULL_PARAMS or
+                # Also disable mixed precision in model eval mode
+                not self._fully_sharded_module.training
```
How should we handle hybrid `eval()` and non-`eval()`? What if the parent module is in `eval()`, but the child module is not?

For a single fully sharded module per handle, taking the fully sharded module's training mode makes sense, but if we ever break that to do execution order wrapping, then what should we do then? Force full precision as long as _any_ of the parent modules is in eval mode?

rohan-varma replied:
I don't see a use case for hybrid eval / non eval. Why would the user want some parts of the model to be in training and others to be eval? Unless they're doing something like disabling dropout / batchnorm tracking for only part of the model.

rohan-varma replied:
If a use case for hybrid eval / non-eval comes up, we can consider additional discussions. Currently, I believe there is always one fully sharded module per handle so this check should be okay until we have execution order based wrapping?

awgu replied:
I agree this is not a major concern right now. I mainly was just exercising the thought :)

I think this should be fine as is.

rohan-varma committed:
Update on "[FSDP] Full precision in eval mode"


If model.eval() is true, then runs the model in full precision.

Changes:
- Changed _force_full_precision to check self.is_training
- Check for _force_full_precision when casting gradients to reduced dtype
- Small change when accessing _full_prec_param_padded
- tests for class based and fully_shard APIs

Differential Revision: [D43933690](https://our.internmc.facebook.com/intern/diff/D43933690/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Full precision in eval mode"


If model.eval() is true, then runs the model in full precision.

Changes:
- Changed _force_full_precision to check self.is_training
- Check for _force_full_precision when casting gradients to reduced dtype
- Small change when accessing _full_prec_param_padded
- tests for class based and fully_shard APIs

Differential Revision: [D43933690](https://our.internmc.facebook.com/intern/diff/D43933690/)

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_mixed_precision.py:
```diff
@@ -165,67 +168,69 @@ def __init__(self, param_dtype, buffer_name="buffer"):
+                        num_active_fsdp += 1
+                        # This FSDP unit is active, verify param points to mixed
+                        cls.assertEqual(param.dtype, expected_param_type)
+                        # _rebuild_full_param should have also freed the fp16 shard.
```
This is not for the current PR, but maybe we should change this:
```suggestion
                        # _unshard should have also freed the fp16 shard.
```
`_unshard()` in `_runtime_utils.py` -> `FlatParamHandle.post_unshard()` should free it

rohan-varma committed:
Update on "[FSDP] Full precision in eval mode"


If model.eval() is true, then runs the model in full precision.

Changes:
- Changed _force_full_precision to check self.is_training
- Check for _force_full_precision when casting gradients to reduced dtype
- Small change when accessing _full_prec_param_padded
- tests for class based and fully_shard APIs

Differential Revision: [D43933690](https://our.internmc.facebook.com/intern/diff/D43933690/)

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -713,6 +759,9 @@ def _post_backward_hook(
             if (
                 not _low_precision_hook_enabled(state)
                 and flat_param.grad.dtype != handle._reduce_dtype
+                # If we are forcing full precision but communicating grads
```
I am a bit confused by the "but communicating grads" part of the comment. IIUC, at this point, we are already certainly communicating gradients, as this is the pre-collective downcast.

Is the statement just meaning like "if we are forcing full precision, don't downcast gradient"?

awgu reviewed (approved):
This looks good to me!

rohan-varma committed:
Update on "[FSDP] Full precision in eval mode"


If model.eval() is true, then runs the model in full precision.

Changes:
- Changed _force_full_precision to check self.is_training
- Check for _force_full_precision when casting gradients to reduced dtype
- Small change when accessing _full_prec_param_padded
- tests for class based and fully_shard APIs

Differential Revision: [D43933690](https://our.internmc.facebook.com/intern/diff/D43933690/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Full precision in eval mode"


If model.eval() is true, then runs the model in full precision.

Changes:
- Changed _force_full_precision to check self.is_training
- Check for _force_full_precision when casting gradients to reduced dtype
- Small change when accessing _full_prec_param_padded
- tests for class based and fully_shard APIs

Differential Revision: [D43933690](https://our.internmc.facebook.com/intern/diff/D43933690/)

[ghstack-poisoned]

rohan-varma commented:
RuntimeError: inductor/test_minifier 1/1 failed!

is unrelated.

rohan-varma commented:
@pytorchbot merge -f "CI done"

rohan-varma committed:
Update on "[FSDP] Full precision in eval mode"


If model.eval() is true, then runs the model in full precision.

Changes:
- Changed _force_full_precision to check self.is_training
- Check for _force_full_precision when casting gradients to reduced dtype
- Small change when accessing _full_prec_param_padded
- tests for class based and fully_shard APIs

Differential Revision: [D43933690](https://our.internmc.facebook.com/intern/diff/D43933690/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge -f "just rebased"

rohan-varma committed:
Update on "[FSDP] Full precision in eval mode"


If model.eval() is true, then runs the model in full precision.

Changes:
- Changed _force_full_precision to check self.is_training
- Check for _force_full_precision when casting gradients to reduced dtype
- Small change when accessing _full_prec_param_padded
- tests for class based and fully_shard APIs

Differential Revision: [D43933690](https://our.internmc.facebook.com/intern/diff/D43933690/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Full precision in eval mode"


If model.eval() is true, then runs the model in full precision.

Changes:
- Changed _force_full_precision to check self.is_training
- Check for _force_full_precision when casting gradients to reduced dtype
- Small change when accessing _full_prec_param_padded
- tests for class based and fully_shard APIs

Differential Revision: [D43933690](https://our.internmc.facebook.com/intern/diff/D43933690/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Full precision in eval mode"


If model.eval() is true, then runs the model in full precision.

Changes:
- Changed _force_full_precision to check self.is_training
- Check for _force_full_precision when casting gradients to reduced dtype
- Small change when accessing _full_prec_param_padded
- tests for class based and fully_shard APIs

Differential Revision: [D43933690](https://our.internmc.facebook.com/intern/diff/D43933690/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge -f "CI done"
